kafka/pykafka/simpleconsumer.py

```diff
@@ -240,18 +239,23 @@ def fetch(self):
         returned messages in the approprate OwnedPartition.
         """
         for broker, owned_partitions in self._partitions_by_leader.iteritems():
-            reqs = [owned_partition.build_fetch_request(self._fetch_message_max_bytes)
-                    for owned_partition in owned_partitions
-                    if owned_partition.message_count < self._queued_max_messages
-                    and owned_partition.empty]
+            reqs = []
+            for owned_partition in owned_partitions:
+                # attempt to acquire lock, just pass if we can't
+                if owned_partition.lock.acquire(False):
+                    has_room = owned_partition.message_count < self._queued_max_messages
+                    if owned_partition.empty and has_room:
+                        reqs.append(owned_partition.build_fetch_request(
+                            self._fetch_message_max_bytes))
             if reqs:
                 response = broker.fetch_messages(
-                    reqs, timeout=self._socket_timeout_ms,
+                    reqs, timeout=self._fetch_wait_max_ms,
                     min_bytes=self._fetch_min_bytes
                 )
                 for partition_id, pres in response.topics[self._topic.name].iteritems():
-                    partition = self._partitions_by_id[partition_id]
-                    partition.enqueue_messages(pres.messages)
+                    owned_partition = self._partitions_by_id[partition_id]
+                    owned_partition.enqueue_messages(pres.messages)
+                    owned_partition.lock.release()
```

Comment:
I've only tested it locally with two threads. Today I'll pump up the thread count and see what's affected. The threading logic is actually really simple as threading logic goes, though. It's just locking each partition to make sure it's ignored if another thread is already dealing with it.

I say it's simple and am probably wrong. :)